refactor(chips): Deprecate md-on-append in favor of new methods.

topherfangio opened the pull request:
The usage of `md-on-append` was not well documented and it's behavior was inconsistent and confusing; many users assumed it was a simple notification of chip additions which caused issues.

After much discussion, we have renamed `md-on-append` to `md-transform-chip` and provided a new `md-on-add` method that is strictly a notification.

Additionally, we have updated the docs and functionality of `md-transform-chip` to show expected return values and their associated behavior.

This new behavior also adds support for simlultaneously using an autocomplete to select an existing value along with the ability to create new chips. The most common case for this is a tag system which shows existing tags, but also allows you to create new ones.

Demos have been updated to show new functionality as well as to workaround a few display issues with the contact chips demo (#4450).

_**Note 1:** This work supercedes PR #3816 which can be closed when this is merged._

_**Note 2:** This also adds an autocomplete demo inside of a dialog to ensure I did not break that functionality._

BREAKING CHANGE:
`md-on-append` has been renamed/deprecated in favor of `md-transform-chip` or the simple notifier `md-on-add`.

We expect to remove this completely in 1.0, so please update your code to use one of the new methods.

Fixes #4666. Fixes #4193. Fixes #4412. Fixes #4863.

topherfangio commented:
@typotter - @ThomasBurleson asked if you could please review this chips PR for me when you have a moment :smile:

typotter commented on src/components/chips/js/chipsDirective.js:
```diff
@@ -67,11 +67,16 @@
    *    the input and delete buttons
-   * @param {expression} md-on-append An expression that when called expects you to return an object
-   *    representation of the chip input string.
+   * @param {expression} md-on-append An expression of form `myFunction($chip)` that when called
```
I think this is not a good API... Partly my fault for the bad naming.
The original intent of md-on-append was to provide a means for the Chips user to build a new object from the input string. It should really have been named something like "chip-builder".
Using the md-on-append as a notification mechanism is troublesome because the String published is often not the final value to be added (esp. since it's expected that on-append will change it), and would be further broken if any other processing is added.

Is md-on-select not sufficient for notifying of new chips?

topherfangio replied:
The `md-on-select` could indeed be used, but that forces the user to watch the list for changes to see whether the item was added or just selected by the user.

Additionally (and more importantly), there is currently no way to use the autocomplete functionality along with the "add chip" functionality, so if users attempt to do this, they wind up getting the chip from the autocomplete AND a blank/empty chip (#4412).

typotter replied:
Would it make more sense then to add a properly named event method for when something is added? like md-on-add? We could mark md-on-append as deprecated, giving it a better name like md-build-chip to avoid confusion.

topherfangio replied:
Okay, talked briefly with Thomas, and we came up with splitting it into `md-transform-chip` and `md-on-append`.

Sound good to you? If so, I'll refactor it soon.

typotter replied:
Sounds Great. Thanks Topher (And Thomas)!

topherfangio commented:
@typotter Okay, I have renamed `md-on-append` as `md-transform-chip`, added the new `md-on-add` method and added a deprecation warning urging users to use one of the new methods.

Can you take one more look and make sure it looks good?

typotter commented on src/components/chips/js/chipsController.js:
```diff
@@ -73,10 +73,32 @@ function MdChipsCtrl ($scope, $mdConstant, $log, $element, $timeout) {
    * @type {boolean}
+   *
+   *
+   * DEPRECATED: Will remove in 1.0.
```
@deprecated ?

typotter commented on src/components/chips/js/chipsController.js:
```diff
@@ -189,30 +216,40 @@ MdChipsCtrl.prototype.getAdjacentChipIndex = function(index) {
-     newChip = this.onAppend({'$chip': newChip});
-   }
+  // If items contains an identical object to newChip, do not append
+  if(angular.isObject(newChip)){
```
nit: space after if

typotter commented on src/components/chips/js/chipsController.js:
```diff
@@ -189,30 +216,40 @@ MdChipsCtrl.prototype.getAdjacentChipIndex = function(index) {
+    }
+  }
 
-   // If useOnAppend and onAppend function is provided call it.
```
shouldn't you still be using onAppend, since it hasn't been removed yet?

typotter commented on src/components/chips/js/chipsController.js:
```diff
@@ -220,9 +257,35 @@ MdChipsCtrl.prototype.getAdjacentChipIndex = function(index) {
 MdChipsCtrl.prototype.useOnAppendExpression = function() {
-  this.useOnAppend = true;
+  this.$log.warn("md-on-append is deprecated; please use md-transform-chip or md-on-add instead");
+  this.useTransformChip = true;
```
maybe only set these if md-transform-chip is not already set.

typotter commented:
Left some comments. Overall LGTM. nice work!

topherfangio commented:
@typotter Thanks! I've made your suggested changes :-)

@ThomasBurleson Should be ready for merge now!
